src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,758 @@
+        {
+            if (ioReadEof(storageReadIo(storageRead)))
+            {
+                THROW(FormatError, "Unexpected WAL end");
+            }
+
+            bufUsedZero(buffer);
+            size = ioRead(storageReadIo(storageRead), buffer);
+            bufUsedSet(buffer, size);
+        }
+    }
+
+    MEM_CONTEXT_TEMP_END();
+    return true;
+}
+
+static void
+walFilterProcess(THIS_VOID, const Buffer *const input, Buffer *const output)
+{
+    THIS(WalFilterState);
+
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(WAL_FILTER, this);
+        FUNCTION_LOG_PARAM(BUFFER, input);
+        FUNCTION_LOG_PARAM(BUFFER, output);
+    FUNCTION_LOG_END();
+
+    // Avoid creating local variables before the record is fully read.
+    // Since if the input buffer is exhausted, we can exit the function.
+
+    if (input == NULL)
+    {
+        // We have an incomplete record at the end
+        if (this->current_step != 0)
+        {
+            const size_t size_on_page = this->got_len;
+
+            // if xl_info and xl_rmid of the header is in current file then read end of record from next file if it exits
+            if (this->got_len >= offsetof(XLogRecord, xl_rmid) + SIZE_OF_STRUCT_MEMBER(XLogRecord, xl_rmid) && getEndOfRecord(this))
+            {
+                filterRecord(this);
+            }
+
+            checkOutputSize(output, size_on_page);
+            memcpy(bufRemainsPtr(output), this->record, size_on_page);
+            bufUsedInc(output, size_on_page);
+        }
+        this->done = true;
+        FUNCTION_LOG_RETURN_VOID();
+        return;
+    }
+
+    if (this->is_begin)
+    {
+        this->is_begin = false;
+        readPage(this, input, 0);
+
+        if (this->current_header->xlp_info & XLP_FIRST_IS_CONTRECORD && !(this->current_header->xlp_info & XLP_FIRST_IS_OVERWRITE_CONTRECORD))
+        {
+            if (readBeginOfRecord(this))
+            {
+                // Remember how much we read from the prev file in order to skip this size when writing.
+                const size_t offset = this->got_len;
+                this->input_offset = 0;
+                lstClearFast(this->headers);
+                if (!readRecord(this, input))
+                {
+                    // Since we are at the very beginning of the file, let's assume that the current input buffer is enough to fully
+                    // read this record.
+                    THROW(FormatError, "record is too big");
+                }
+                filterRecord(this);
+
+                memmove(this->record, ((char *) this->record) + offset, this->record->xl_tot_len - offset);
+                this->got_len -= offset;
+
+                writeRecord(this, output);
+                this->i++;
+                this->same_input = true;
+                lstClearFast(this->headers);
+                FUNCTION_LOG_RETURN_VOID();
+                return;
+            }
+            else
```

Comment:
Here the else statement is redundant, since the if statement above ends with a return.